active/0000-engines.md

```diff
@@ -0,0 +1,411 @@
+  routing maps of parent applications or engines. Routing maps are alway eager
+  loaded, which allows for deep linking into an engine's routes regardless of
+  whether the engine itself has been instantiated.
+
+* Route-less engines can isolate complex functionality that is not related to
+  routing (e.g. a chat engine in a sidebar). Route-less engines can be rendered
+  into outlets ad hoc as routes are loaded.
+
+The potential scope of engines is large enough that this feature merits
+development and delivery in multiple phases. A minimum viable version could be
+released sooner, which could be augmented with more advanced features later.
+
+An initial release of engines could provide the following benefits:
+
+* Distributed development - Engines can be developed and tested in isolation
+  within their own Ember CLI projects and included by applications or other
+  engines. Engines can be packaged and released as addons themselves.
+
+* Integrated routing - Support for mounting routable engines in the routing maps
+  of applications or other engines.
+
+* Ad hoc embedding - Support for embedding route-less engines in outlets as
+  needed.
+
+* Clean boundaries - An engine can cooperate with its parents through a few
+  explicit interfaces. Beyond these interfaces, engines and applications are
+  isolated.
+
+Subsequent releases of engines could allow for the following:
+
+* Lazy loading - An engine could allow its parent to boot with only its routing
+  map loaded. The rest of the engine could be loaded only as required (i.e.
+  when a route in an engine is visited). This would allow applications to boot
+  faster and limit their memory consumption.
+
+* Namespaced access to engine resources from applications - This could open up
+  the potential for applications to use, and extend, an engine's resources much
+  like resources in other addons, but without the possibility of namespace
+  collisions.
+
+## Detailed design
+
+Engines are very similar to regular applications: they can be developed in
+isolation in Ember CLI, include addons, and contain all the same elements,
+including routes, components, initializers, etc. The primary differences are
+that an engine does not boot itself and an engine does not control the router.
+
+### Engine internals
+
+New `Engine` and `EngineInstance` classes will be introduced.
+
+Applications and engines will share ancestry. It remains TBD whether
+applications will subclass engines, or whether a common ancestor will be
+introduced.
+
+Engines and applications will share the same pattern for registry / container
+ownership and encapsulation. Both will also have initializers and instance
+initializers.
+
+Engine instances will have access to their parent instances. An engine's parent
+could be either an application or engine.
+
+#### Routable vs. route-less engines
+
+Routable engines will define their routes in a new `Ember.Routes` class. This
```

Comment:
> is the difference between a routeless and route only the fact that they declare routes or not?

Pretty much. Routable engines can be mounted directly in a routing map. Routeless enginess need to be mounted in an outlet and have no routing concerns.

> Is it possible to boot a routed engine with a location:none, so it doesn't interfere?

Seems possible - any routes in the engine's routing map would then be ignored.

> QP are global, how do we namespace them?

Good question. Since the router now has the potential to be in charge of multiple routing maps, perhaps QP should be scoped to routing maps instead of being global.

> There also needs to be mention on how one links between sync/async engines.

A mounted engine's routing map will be loaded eagerly by the main router. This will enable the router to create links eagerly, even before the engine is instantiated.

Transitions to those links will of course be async and would require instantiating (and perhaps lazy loading) the engine. The instantiation of a mounted engine should be part of the async resolution of a route.